packages/smart-contracts/src/contracts/ERC20TransferrableReceivable.sol

```diff
@@ -0,0 +1,173 @@
+import '@openzeppelin/contracts/token/ERC721/extensions/ERC721URIStorage.sol';
+
+contract ERC20TransferrableReceivable is ERC721URIStorage {
+  using SafeERC20 for IERC20;
+  using Counters for Counters.Counter;
+
+  // Counter for uniquely identifying payments
+  Counters.Counter private _paymentId;
+
+  // Counter for uniquely identifying receivable tokens
+  Counters.Counter private _receivableTokenId;
+
+  // ERC20 that one ERC20TransferrableReceivable is paid in
+  mapping(uint256 => address) public assetToken;
+
+  // Nested mapping for looking up receivable token given a paymentReference
+  // and original payment address
+  mapping(bytes => mapping(address => uint256)) private _receivableTokenIdMapping;
+
+  // Helper mappings to lookup all receivable tokens owned by an address
+  mapping(address => uint256[]) private _receivableIds;
+  mapping(uint256 => uint256) private _receivableIdIndexes;
+
+  address public paymentProxy;
+
+  // Event to declare payments to a receivableTokenId
+  event Payment(
+    address sender,
+    address recipient,
+    address paymentProxy,
+    uint256 receivableTokenId,
+    address assetAddress,
+    uint256 amount,
+    uint256 paymentId,
+    bytes indexed paymentReference,
+    uint256 feeAmount,
+    address feeAddress
+  );
+
+  // Event to declare a transfer with a reference
+  // This event is emitted from a delegatecall to an ERC20FeeProxy contract
+  event TransferWithReferenceAndFee(
+    address tokenAddress,
+    address to,
+    uint256 amount,
+    bytes indexed paymentReference,
+    uint256 feeAmount,
+    address feeAddress
+  );
+
+  constructor(
+    string memory name,
+    string memory symbol,
+    address _paymentProxyAddress
+  ) ERC721(name, symbol) {
+    paymentProxy = _paymentProxyAddress;
+  }
+
+  function supportsInterface(bytes4 interfaceId) public view override(ERC721) returns (bool) {
+    return super.supportsInterface(interfaceId);
+  }
+
+  function payOwner(
+    uint256 receivableTokenId,
+    uint256 amount,
+    bytes calldata paymentReference,
+    uint256 feeAmount,
+    address feeAddress
+  ) external {
+    require(amount != 0, 'Zero amount provided');
+    address owner = ownerOf(receivableTokenId);
+    _paymentId.increment();
+    address assetAddress = assetToken[receivableTokenId];
+
+    (bool status, ) = paymentProxy.delegatecall(
+      abi.encodeWithSignature(
+        'transferFromWithReferenceAndFee(address,address,uint256,bytes,uint256,address)',
+        assetAddress,
+        owner,
+        amount,
+        paymentReference,
+        feeAmount,
+        feeAddress
+      )
+    );
+    require(status, 'transferFromWithReferenceAndFee failed');
+
+    emit Payment(
+      msg.sender,
+      owner,
+      paymentProxy,
+      receivableTokenId,
+      assetAddress,
+      amount,
+      _paymentId.current(),
+      paymentReference,
+      feeAmount,
+      feeAddress
+    );
+  }
+
+  function mint(
```

Comment:
I think there are drawbacks as well, but if we saved the total amount on mint, there are a few advantages:
1/ once the balance is complete, we can burn the token and avoid (mistaken) double payments.
2/ one can trade a partially paid receivable